### ISSUE TITLE
Output Audio?

### ISSUE THREAD

**Issue by linuxmobile** [alemidev/scope-tui#4]
First of all, good morning, thank you for the tool, it is very nice and interesting.

Following this, I wanted to ask you why you only capture the input audio.  Contrary to what cava does.

I'm working on creating the flake so people can install/use it on NixOS, and then do the PR for you, but I couldn't really make sense of it only capturing the input audio, and not the output.

**Comment by alemidev** [comment]
hi! thanks for your interest

`scope-tui` doesn't pick any specific audio device and works with both input and output sources! upon starting it up, you can pass an audio device as argument: for example, I use `scope-tui pulse pipewire.monitor` to attach to my master audio monitor.

when started without arguments, it will choose your default audio device, which usually is your mic

it's possible to re-route audio at runtime freely, for example i use [Helvum](https://gitlab.freedesktop.org/pipewire/helvum) to connect different input channels to different output sources and compare their waveforms

currently it only supports pulseaudio because its Simple Api is, well, easy! but it's not really super friendly to work with, i'm probably switching to [cpal](https://github.com/RustAudio/cpal) soon:tm: and then i may add some better controls to manage inputs directly from scope-tui

**Comment by linuxmobile** [comment]
Sorry for responding too late. Even adding the monitor with "pipewire.monitor" I can't get it to work!

![image](https://github.com/alemidev/scope-tui/assets/10554636/7ced09f3-90d8-41f3-9554-f4c4af55751f)

**Comment by alemidev** [comment]
hey! i'm also keen on responding late, no worries and thanks for waiting! (: i have some solutions to propose:

---

the `audio` source doesn't yet support choosing which device it should get attached to. if your system has pipewire and pipewire-pulse you can compile scope-tui with the `pulseaudio` feature flag (`cargo build --release --features=pulseaudio`) and then change `audio` for `pulse` in your command:

`$ ./scope-tui pulse pipewire.monitor`

this is kind of a workaround, the `pulseaudio` is not great, but it will allow you to control the mapping directly from scope-tui

---

you can rewire scope-tui audio manually. i'm not sure if there's a simple way to do it on bare pulseaudio (maybe with `pavucontrol`?) but if you're running the pipewire audio server you can use `helvum` to rewire audio accordingly: first disconnect whatever scope-tui attached to by default (by dragging the wires), and then connect it to your main audio monitor (again by dragging). iirc there's a ~60s timeout on CPAL, so you have to be not too slow otherwise scope-tui will close!

---

last is the real solution: waiting (sorry!). the `audio` source is still experimental, i'm trying to get it to feature parity with the pulseaudio source (to support multiple platforms and setups) but windows is giving me a huge headache! it doesn't respect my buffer size, but i digress. right now i'm very focused on another project of mine which really caught my interest. it's starting to fade away and in the upcoming month i'd like to look back into scope-tui, and get this feature properly adapted

---

so TLDR, either:
 * temporarily switch to `pulseaudio` source
 * temporarily re-wire scope-tui sources inside your audio server config
 * wait for me to fix the `audio` source
 
sorry for the inconvenience and thanks for your interest!

**Comment by alemidev** [comment]
Hi @linuxmobile , this is a rather old issue but I think I kind of solved this with the latest release: I got around to implementing the `audio` source with `cpal` which should be more user friendly.

Now `scope-tui` can hook to any of your sound system input devices!

To be able to listen to the output audio, your system needs to have a loopback/monitor device. On Linux, pipewire/pulseaudio provide the "monitor" inputs, which just spit out your output audio.

MacOS for example doesn't have any monitor device out of the box, and external software needs to be employed (for example, [BlackHole](https://github.com/ExistentialAudio/BlackHole))

I'm not sure about windows

So basically, I don't think I can circumvent system limits from inside scope-tui: if no monitor device is provided as input, scope-tui will only be able to attach to your microphone, or other inputs. When monitor inputs are provided, they should be usable by scope-tui with the `audio` source. You should be able to list all available inputs with `scope-tui audio -l`.

I think this issue can be closed, but feel free to reopen if you have more feedback or issues!
